### ISSUE TITLE
Translating Exercism to other languages

### ISSUE THREAD

**Issue by Rachc** [exercism/exercism#4207]
Hello

I was talking to some friends (@woliveiras and @laurenmariaferreira) and we would love to translate exercism to pt-br. We want to reach a larger audience that does not speak English (: 
I'm opening this issue to know if I can do this and, if it's positive, to getter some help

Thank you

**Comment by wesmelo** [comment]
i can help with the translation 👊

**Comment by dreecarvalho** [comment]
I'd be glad to help to translate if needed :)

**Comment by anapaulagomes** [comment]
Me too. :)

**Comment by iHiD** [comment]
Katrina, Nicole and myself are going to have a proper chat about this and
come back with some structured thoughts. In all likelihood that won't be
for a few days though. In the meantime thank you for all offering to help
out. You're all awesome :)

**Comment by pehprado** [comment]
I can help to translate 😃

**Comment by rodolfolottin** [comment]
count on me!

**Comment by shadowmaru** [comment]
I'll be able to help 😄

**Comment by magodoy88** [comment]
I can help too! :)

**Comment by davibusanello** [comment]
I make myself available too

**Comment by Rachc** [comment]
@magodoy88 @rodolfolottin @anapaulagomes @pehprado @davibusanello
While @iHiD talk with the staff about this, We can organize ourselves via telegram.
Please, contact me on twitter ( `@_rachcl` ) so I can pass the link through DM (:

**Comment by thais** [comment]
I can help 🐗

**Comment by douglasgimli** [comment]
I'm on it too!

**Comment by weltonsanches** [comment]
I can help too :)

**Comment by Rachc** [comment]
Hello
I'm just pinging on this issue (: 
We are still very excited to do this and we are waiting for your return.
Thank you

**Comment by iHiD** [comment]
@Rachc @douglasgimli (and everyone else) Thanks for your enthusiasm and nudges. I want to just emphasise again that achieving this will be hugely complicated. There would be well over 10,000 files that would need translating, and then a plan for how we can move forward without being blocked by translations for future changes. It's going to take us a bit of time to work out if this is a project we feel is something we have the resource to cope with or manage on an ongoing basis atm. We will get back to you with some proper thoughts but ask for your patience as we work through some of the more urgent bugs and issues first. Thanks :)

**Comment by douglasgimli** [comment]
Great @iHiD, thanks for the feedback, we'll sit tight and wait for new instructions! :)

**Comment by iHiD** [comment]
Hey everyone. So we've spent some time discussing this today. 

I think the first key thing is to be clear that right now, Exercism provides mentoring only in English. Until we can keep up with the amount of submissions we're getting in English, we're not going to expand mentoring to other languages, because we simply won't cope with the demand. In the long-run this will change, but not in the next 6 months.

So what that in mind, the key thing we need is to get a better understanding of which parts of Exercism would improve because of translation and why? If (for now) English is a prerequisite to be mentored, what value will translating parts of the website into a different language give our users, and which bits are the key pain points?

If you could explain that to us a bit more we will hopefully be able to find a way forward :)

**Comment by pernambucano** [comment]
Another Brazilian here. Would love to help translate exercism as well. 
@iHiD , I guess we don't need to focus on the mentoring part right now. Lot's of Brazilians know how to read in English pretty well, but feel intimidated by a whole English only website. To narrow down even further I'd say we start by only translating the common pieces of advice/tutorials. It'll help those who at least want to get started.  Another thing to think about is that the majority of universities here use just a few languages in their 101 courses (mainly Java, Python, and C/C++.)

**Comment by Rachc** [comment]
@iHiD We were talking with some of the people the comment on this topic about our motivation to translate exercism to PT-BR (since you are with one already have a very high demand now for students).
Brazil is a big country and many people don't know a second language. Most people that want to work with IT don't know English (yet), and have difficulty accessing good study material. Initiatives like exercism and FreeCodeCamp doesn't exist, which makes more difficult for us.

I know that exercism objective isn't to teach people how to code, but I think you are excellent in making a person more comfortable with a programming language that they are learning (and, consequently, help they fell more prepared to apply for a new job opportunity)

We have a group in Brazil called Training Center that has a lot of people willing to share knowledge, and I believe that if we translate exercism, a lot of these people could mentor too (but I know that student demand is likely to be higher than mentors)

I think the most crucial point to translate is the exercise readme and - the most complicated - It will be interesting if the review of the exercises could be done in Portuguese (maybe signaling that the person who did the exercise did it in PT? And we have to think in a way to differentiate the exercise in PT or EN when you download. Maybe another CLI?)

The most difficult point, in my opinion, is to know there are more than 10 000 files to translate, but if we accept that the PT version will be partially translated for a while, we can translate gradually without impacting the development of the rest of the exercism (:

**Comment by Rachc** [comment]
And for the Portuguese speakers, we are organizing ourselves in the telegram, but we did not translate anything, because we need to know first if we can (: 
(you can send me a DM on twitter `_rachcl` and if we have authorization to translate, I pass you the group link)

**Comment by davibusanello** [comment]
@Rachc I make these words my own.
My sample, when I started in IT I could badly read some words in English and I was limited by learning with low-quality content and in the wrong way.
I believe that making this great content more accessible to the Portuguese speakers' newcomers or mentors in any programing language, goes meeting the purpose of the Exercism

**Comment by douglasgimli** [comment]
Yes, I agree! We Brazilians live in a different environment, and the majority part (our junior developers) are scary by a content in others languages. Imagine yourself, is hard enough to start coding, what if all of the content you need is in a foreign language – and believe we don't have a second default language here.
If we can help our beginners' developers to start and/or make their first step easy somehow – like giving them an accessible content – this will just help our community to grow.

**Comment by iHiD** [comment]
The question in still unsure of is that if the mentoring is in English, rhe
regardless of whether the site is in English, will the beginner still find
the product scary and not use it. Are we not lulling them into a false
sense of security?

**Comment by pernambucano** [comment]
Hi, @iHiD. I guess one possibility is to focus on a limited part of exercism. As a starting point we could limit the user to the mentoring part or tell him that we only have mentoring in English. All in all, there's this huge part of exercism that would still be available to any portuguese speaker.

**Comment by nunocf** [comment]
I'm from Portugal, not Brasil, and I'd be happy to give you guys some help if you ever decide to translate to PT-PT. The website is great and I can see it being widely used in all of Portugal's universities.

Once I get more versed in exercism, I'd love to be able to become a mentor and help some people in my own language (PT-PT or PT-BR, or even english, if necessary).

For those that don't know, PT-PT and PT-BR are just slightly different, but seeing that the effort for PT-BR is being made, tweaking it to PT-PT is a very quick effort.

Let me know your thoughts!

**Comment by mattastrophe** [comment]
just throwing in some info for this discussion.  We've been using OmegaT (https://omegat.org) for managing our translations (and translation memory) for a few years.  It's worth considering as part of how you manage the translations (and the memories can be stored in git for future/multiple translators).

**Comment by trotro** [comment]
I'd like to help translating exercism into french also :)
On another website they're using [pootle](https://pootle.translatehouse.org/index.html) to manage the translation by the community. Maybe it's worth looking.

**Comment by cschappert** [comment]
Hi @iHiD !
I'm an American living in Tokyo and, wondering if there was a way I could contribute to a Japanese translation, found this issue.

I'm actually just now on v2 and learning how the teams edition relates to the personal edition. My understanding is that the teams edition does not include mentoring and that solutions are shared only with team members. (Team members are therefore responsible for mentoring each other.)

Reflecting on what I've learned watching developers who don't speak or read English using tools and services in English, I'm thinking the following:

If translating the product is to be done incrementally, the best place to start is the exercise readme's. The translations should only be made available on the teams edition. This way companies, communities, etc. could create a group, read the exercise descriptions in their native language and receive mentoring in their native language by other people on their team / in their community.

I think having a translated landing page for the teams edition that basically lets the user know that the translation is a work in progress and how they are expected to interact with the product (at least for the time being) would also be helpful.

It's pretty common practice for people here to write Japanese language blog posts explaining how to get started using popular English language services (Here's one for v1 of exercism.io of which I am **not** the author: https://qiita.com/Saayaman/items/4d1cc2b77fe48704ddf0 [all credit goes to the original author]).

Lastly, I have to say that I do have some mixed feelings about doing translations. While having services like Stack Overflow in Japanese has given non-English speaking developers a place to gather and share information, having separate communities like this can also create something like an information silo. That said, coming up with a common language for the entire world to share information in is probably out of scope for this issue :sweat_smile: 

While I've spoken mostly on the topic of developers here in Tokyo, I hope that this is useful information re: translation in general.

If the project does get moving, I'd be happy to rally a team together to start on the JP translation.

**Comment by sshine** [comment]
This comment is intended to summarize the subject of translating Exercism.

Most comments are made by people who are willing to contribute translations. This was followed by a discussion about the purpose of translating the site when mentoring is still English-only. While there is still no conclusions made, the status is that translating the website is not a priority yet.

This issue is left open to continue further discussion.

A quick summary of open ends in the discussion from where it ended last year:

@iHiD wrote:

> Thanks for your enthusiasm and nudges. I want to just emphasise again that achieving this will be hugely complicated. There would be well over 10,000 files that would need translating, and then a plan for how we can move forward without being blocked by translations for future changes. [...]

> [...] the key thing we need is to get a better understanding of which parts of Exercism would improve because of translation and why? If (for now) English is a prerequisite to be mentored, what value will translating parts of the website into a different language give our users, and which bits are the key pain points?

> The question in still unsure of is that if the mentoring is in English, rhe regardless of whether the site is in English, will the beginner still find the product scary and not use it. Are we not lulling them into a false sense of security?

@pernambucano wrote:

> Lot's of Brazilians know how to read in English pretty well, but feel intimidated by a whole English-only website.

> *[...]* one possibility is to focus on a limited part of exercism.
> *[...]* tell *[the user]* that we only have mentoring in English.

@mattastrophe wrote:

> We've been using OmegaT (https://omegat.org) for managing our translations (and translation memory) for a few years. [...]

@cschappert suggests (see [full comment](https://github.com/exercism/exercism/issues/4207#issuecomment-487339568)) that **a translated Exercism Teams may be valuable to non-English speakers in spite of English-only mentoring on Exercism**. He also says:

> Lastly, I have to say that I do have some mixed feelings about doing translations. While having services like Stack Overflow in Japanese has given non-English speaking developers a place to gather and share information, having separate communities like this can also create something like an information silo.

*(I have personally wondered about the language-based fragmentation on the internet and thought that maybe the silo view is Americentric. With the [.рф TLD](https://en.wikipedia.org/wiki/Internationalized_domain_name), did we create a silo or give an existing silo a name that some people can't pronounce?)*

**Comment by cschappert** [comment]
@sshine Thank you for summarizing everything!

# TL;DR:
I am happy to help growing and strengthening an international OSS community that every developer can participate in without being advantaged or disadvantaged unfairly by their language abilities. One way (just throwing out ideas here) to accomplish this might be community translations, another might just be abandoning the idea of a de facto lingua franca and learning each other's languages (the least practical, but maybe the most rewarding?). One more might be (in the case of an English site) at least making sure that the English is as clear and understandable as possible from the perspective of a non-native speaker. (Sci-Fi backstories a la Advent of Code are fun and I love them but they make it *really* hard for many non-native speakers. Really good UI / UX can make language a non-issue sometimes, and we all speak code so putting the code front and center is possibly another way to break down the language barrier.

# The first thing that popped into my head

> Lastly, I have to say that I do have some mixed feelings about doing translations. While having services like Stack Overflow in Japanese has given non-English speaking developers a place to gather and share information, having separate communities like this can also create something like an information silo.

It took me a little while to remember why I said this because it was probably a little off-topic and the context is missing.

At the time I was working at a Japanese company at which I was the only non-Japanese person on the development team and the only member comfortable with using English at work (I'm not just comfortable, it's my native language). I was in charge of on-boarding and education among other things, and I spent a lot of time searching for something like exercism.io, but in Japanese. There are of course services, but in my opinion we just don't have the variety and volume of educational resources that the English-speaking (native or otherwise) development community has. Not all, but much of the material available is translated and that means it's sometimes out of date as well.

One obvious solution is that developers here (wherever they might be from) develop these resources, but there's sort of a catch-22 because in order to know enough about a non-local technology (i.e. not originally made here) to teach it as well as a local would there needs to be again a sort of interpreter in-between the two communities. The most obvious example of a local scene here is of course Ruby and it continues to be a really active community because we don't have the bottleneck that occurs when we're trying to interpret from primary sources. (I say "we" here as I don't have to wait for translations, but I do often need to create them for team-members).

OTOH, I hope I'm only seeing what's in my field of vision and simply over-estimating the volume of English language resources in comparison to other languages. In a way, I hope I am!

Part of me feels very uncomfortable telling people to "just learn English", but it does seem to be a huge advantage for developers and creates tons of opportunities to be a part of an international OSS community. I am fluent in Japanese and I promise I'm not trying to get my co-workers to speak English just because I'm lazy. :smile: There's also a pretty big push here coming from bilingual Japanese developers to get their team comfortable with English and reading primary sources. In a way, every time I write a translation for someone, I prevent this. 

Whether we should have one global community or lots of different "centers" that interact with each other / whether there should be a de facto lingua franca and whether it should be English I don't know. If you are interested in this subject please contact me privately as it's a subject I'm always happy to discuss and I'm probably drifting way off-topic for this specific issue. Personally I love the diversity of the open source community and I'm happy to be a part of any project that can help us all share ideas, work together and learn from each other as equals (i.e. not advantaged or disadvantaged by being a native or non-native speaker of X)

**Comment by iHiD** [comment]
> One more might be (in the case of an English site) at least making sure that the English is as clear and understandable as possible from the perspective of a non-native speaker. 

I agree a lot with this. I [PR'd a while back](https://github.com/exercism/problem-specifications/pull/1360) when I did hamming and felt the language would be very difficult for non-native speakers.

I'd happily welcome suggestions to where we can simplify the language on the site, and PRs to exercism/website-copy or exercism/problem-specifications repos would be even more amazing.

---

I recently wrote the following on our Slack room to give an official position in a similar discussion. Maybe it will add shed some light on where I feel we're at:

> So we crave to be a place where people can come and learn, whatever their background. And that means whatever language they speak. I believe, if presented as part of a journey (ie learn to code, learn to communicate around code, find a (remote?) job), exercism can help someone get from a very uncomfortable work-situation to a more lucrative one, and therefore offer a method of social mobility.
> 
> However, we have to get there. Adding foreign language support opens huge floodgates that make Exercism less sustainable. Right now, we struggle to cope with demand just from English-speaking students. Offering support to others would overwhelm us more, and I think therefore it's important to manage expectations around that. For example, I love the idea of having flags next to people to say what language they speak, but it would add an expectation that someone would then get paired based on that. That's not technically difficult, but with so few active mentors, we would likely see queues flooded with people waiting for mentoring in their native language, and no-one to mentor them. Which then actually means their experience is worse because they don't just take the English -> Google Translate route, which does work for many languages (albeit less so for eastern ones). There are other things too, like how do we deal with Code of Conduct violations in languages we don't speak. How do we offer help with installation issues in different languages. The moment we start to offer "some" support, we will be inundated different challenges we'll have to work out.
> 
> So I passionately agree that this is where I want Exercism to get to (it's one of the primary reasons I do Exercism!), but it's also my job to make sure Exercism gets anywhere, and so right now, to not overwhelm us, we need to get things to a point that are sustainable and work in English, and then we can start to offer features that help people in other languages - somewhere I'm very excited to get to.

**Comment by martl** [comment]
I can help translate English ->> German

**Comment by Airbus5717** [comment]
I can Help English <-> Arabic 
Ping me when you need anything

**Comment by e5pe** [comment]
I can help translate to Spanish. Let me know if you need me.

**Comment by ErikSchierboom** [comment]
Hello 👋

With the launch of Exercism v3, we are **closing all issues in this repository** to help give us a clean slate to detect new problems. If this issue is still relevant to Exercism v3 (e.g. it's a feature that we haven't implemented in v3, or a bug that still exists), please reopen it and we will review it and post an update on it as soon as we get chance.

Thanks for helping make Exercism better, and we hope you enjoy v3 🙂

**Comment by Ben-Taarit** [comment]
I can do the En-Fr translation as well
